File: frontend/web_ccadmin_ecommerce/src/app/enterprise/sale/model/dto/PresaleRegisterDto.ts
import { PresaleChannelEntity } from '../entity/PresaleChannelEntity';
import { PresaleDetEntity } from '../entity/PresaleDetEntity';
import { PresaleHeadEntity } from '../entity/PresaleHeadEntity';

export class PresaleRegisterDto {
  public Headboard: PresaleHeadEntity = new PresaleHeadEntity();
  public DetailList: PresaleDetEntity[] = [];
  public PresaleChannel: PresaleChannelEntity = new PresaleChannelEntity();
  public CreditNoteCod: string = '';

  public rebuild(): void {
    this.Headboard.NumPriceSubTotal = this.money(this.DetailList.reduce(
      (total, item) => total + item.NumUnitPrice * item.NumUnit,
      0
    ));
    this.Headboard.NumDiscount = this.money(this.DetailList.reduce(
      (total, item) => total + item.NumDiscount * item.NumUnit,
      0
    ));
    this.Headboard.NumTotalPrice = this.money(
      this.Headboard.NumPriceSubTotal - this.Headboard.NumDiscount
    );
  }

  private money(value: number): number {
    return Math.round((Number(value || 0) + Number.EPSILON) * 100) / 100;
  }
}

File: backend/ws_wa_store_ccadmin/src/main/java/com/ccadmin/app/sale/model/dto/PresaleRegisterDto.java
package com.ccadmin.app.sale.model.dto;

import com.ccadmin.app.sale.model.entity.PresaleDetEntity;
import com.ccadmin.app.sale.model.entity.PresaleChannelEntity;
import com.ccadmin.app.sale.model.entity.PresaleHeadEntity;
import com.ccadmin.app.sale.model.entity.SaleBillingEntity;

import java.util.List;

public class PresaleRegisterDto {

    public PresaleHeadEntity Headboard;

    public List<PresaleDetEntity> DetailList;

    public PresaleChannelEntity PresaleChannel;

    public SaleBillingEntity SaleBilling;

    public String CreditNoteCod;

}
